pkgs/development/interpreters/python/fetch-python-requirements.nix

```
@@ -0,0 +1,181 @@
+        "aarch64-linux" = [
+          "manylinux1_aarch64"
+          "manylinux2010_aarch64"
+          "manylinux2014_aarch64"
+        ];
+      };
+      platforms = if sysToPlatforms ? "${targetSystem}" then sysToPlatforms."${targetSystem}" else throw ''
+        'binaryOnly' fetching is currently not supported for target ${targetSystem}.
+        You could set 'binaryOnly = false' and execute the build on a ${targetSystem}.
+      '';
+
+      pythonAttrName = lib.stringAsChars (x: if x == "." then "" else x) python.libPrefix;
+
+      # We need to execute pip from the python version for which we want to download packages.
+      # Pip accepts '--python-version', but this works only for wheel packages.
+      pythonPip = buildPackages."${pythonAttrName}".withPackages (ps: [
+        ps.pip
+        ps.setuptools
+        ps.pkgconfig
+      ]);
+
+      # Normalize package names:
+      #   "Example_Package[extra]" -> "example-package"
+      normalizedNames = map (req:
+        lib.toLower (lib.replaceStrings [ "_" ] [ "-" ] (lib.elemAt (lib.splitString "[" req) 0))
+      ) requirements;
+
+      # fixed output derivation containing packages,
+      # each being symlinked from it's normalized name
+      # Example:
+      #   "$out/werkzeug" will point to "$out/Werkzeug-0.14.1-py2.py3-none-any.whl"
+      self = stdenv.mkDerivation {
+
+        name = "python-sources-${targetSystem}";
+
+        outputHashMode = "recursive";
+        outputHashAlgo = "sha256";
+        inherit outputHash;
+
+        nativeBuildInputs = with python3.pkgs; [ pythonWithMitmproxy pythonPip curl cacert ];
+
+        dontUnpack = true;
+        dontInstall = true;
+        dontFixup = true;
+
+        buildPhase = ''
+          # the script.py will read this date
+          export MAX_DATE=${builtins.toString maxDate}
+          pretty=$(python -c '
+          import os; import dateutil.parser;
+          try:
+            print(int(os.getenv("MAX_DATE")))
+          except ValueError:
+            print(dateutil.parser.parse(os.getenv("MAX_DATE")))
+          ')
+          echo "selected maximum release date for python packages: $pretty"
+
+          # find free port for proxy
+          proxyPort=$(python -c '\
+          import socket
+          s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+          s.bind(("", 0))
+          print(s.getsockname()[1])
+          s.close()')
+
+          # start proxy to filter pypi responses
+          # mitmproxy wants HOME set
+          # mitmdump == mitmproxy without GUI
+          HOME=$(pwd) mitmdump \
+            --listen-port "$proxyPort" \
+            --ignore-hosts '.*files.pythonhosted.org.*'\
+            --script ${./filter-pypi-responses.py} &
+
+          # wait for proxy to come up
+          while sleep 0.5; do
+            timeout 5 curl --fail --proxy http://localhost:$proxyPort http://pypi.org 2>/dev/null && break
```

```suggestion
            timeout 5 curl -fLsS --proxy http://localhost:$proxyPort http://pypi.org && break
```

Just a suggestion:

I use those args in many scripts and they are a good mixture between being silent, compatible and still showing important errors. Feel free to adapt.

-f -> --fail
-L -> follow redirects
-s -> silen
-S -> show errors in silent

---

Thanks for that. I tried this and it still seems to print messages like
`curl: (7) Failed to connect to localhost port 37467: Connection refused`.
Do we want this in the build log? Maybe it will confuse people to think something is wrong, even though this is expected.

---

Oh, I should read more carefully. Just needed to remove -S

---

kept only the -fs now, since -L somehow leads to a situation where SSL is required.